// @azure/cognitiveservices-luis-runtime/esm/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
export var appId = {
    parameterPath: "appId",
    mapper: {
        required: true,
        serializedName: "appId",
        type: {
            name: "String"
        }
    }
};
export var bingSpellCheckSubscriptionKey = {
    parameterPath: [
        "options",
        "bingSpellCheckSubscriptionKey"
    ],
    mapper: {
        serializedName: "bing-spell-check-subscription-key",
        type: {
            name: "String"
        }
    }
};
export var endpoint = {
    parameterPath: "endpoint",
    mapper: {
        required: true,
        serializedName: "Endpoint",
        defaultValue: '',
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export var log = {
    parameterPath: [
        "options",
        "log"
    ],
    mapper: {
        serializedName: "log",
        type: {
            name: "Boolean"
        }
    }
};
export var spellCheck = {
    parameterPath: [
        "options",
        "spellCheck"
    ],
    mapper: {
        serializedName: "spellCheck",
        type: {
            name: "Boolean"
        }
    }
};
export var staging = {
    parameterPath: [
        "options",
        "staging"
    ],
    mapper: {
        serializedName: "staging",
        type: {
            name: "Boolean"
        }
    }
};
export var timezoneOffset = {
    parameterPath: [
        "options",
        "timezoneOffset"
    ],
    mapper: {
        serializedName: "timezoneOffset",
        type: {
            name: "Number"
        }
    }
};
export var verbose = {
    parameterPath: [
        "options",
        "verbose"
    ],
    mapper: {
        serializedName: "verbose",
        type: {
            name: "Boolean"
        }
    }
};
