// typedoc-plugin-markdown/dist/theme.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const Handlebars = require("handlebars");
const typedoc_1 = require("typedoc");
const models_1 = require("typedoc/dist/lib/models");
const events_1 = require("typedoc/dist/lib/output/events");
const theme_1 = require("typedoc/dist/lib/output/theme");
const comments_component_1 = require("./components/comments.component");
const utils_component_1 = require("./components/utils.component");
class MarkdownTheme extends theme_1.Theme {
    constructor(renderer, basePath) {
        super(renderer, basePath);
        this.listenTo(renderer, events_1.PageEvent.END, this.onPageEnd, 1024);
        renderer.removeComponent('assets');
        renderer.removeComponent('javascript-index');
        renderer.removeComponent('toc');
        renderer.removeComponent('pretty-print');
        renderer.removeComponent('marked-links');
        renderer.removeComponent('legend');
        renderer.removeComponent('navigation');
        renderer.addComponent('comments', new comments_component_1.CommentsComponent(renderer));
        renderer.addComponent('utils', new utils_component_1.HelperUtilsComponent(renderer));
    }
    static formatContents(contents) {
        return (contents
            .replace(/[\r\n]{3,}/g, '\n\n')
            .replace(/!spaces/g, '')
            .replace(/^\s+|\s+$/g, '') + '\n');
    }
    isOutputDirectory(outputDirectory) {
        let isOutputDirectory = true;
        const listings = fs.readdirSync(outputDirectory);
        listings.forEach((listing) => {
            if (!this.allowedDirectoryListings().includes(listing)) {
                isOutputDirectory = false;
                return;
            }
        });
        return isOutputDirectory;
    }
    allowedDirectoryListings() {
        return [
            this.entryFile,
            this.globalsFile,
            ...this.mappings.map((mapping) => mapping.directory),
            'media',
            '.DS_Store',
        ];
    }
    getUrls(project) {
        const urls = [];
        const entryPoint = this.getEntryPoint(project);
        if (this.readme === 'none') {
            entryPoint.url = this.entryFile;
            urls.push(new typedoc_1.UrlMapping(this.entryFile, Object.assign({}, entryPoint), 'reflection.hbs'));
        }
        else {
            entryPoint.url = this.globalsFile;
            urls.push(new typedoc_1.UrlMapping(this.globalsFile, entryPoint, 'reflection.hbs'));
            urls.push(new typedoc_1.UrlMapping(this.entryFile, project, 'index.hbs'));
        }
        if (entryPoint.children) {
            entryPoint.children.forEach((child) => {
                if (child instanceof typedoc_1.DeclarationReflection) {
                    this.buildUrls(child, urls);
                }
            });
        }
        return urls;
    }
    buildUrls(reflection, urls) {
        const mapping = this.mappings.find((mapping) => reflection.kindOf(mapping.kind));
        if (mapping) {
            if (!reflection.url || !MarkdownTheme.URL_PREFIX.test(reflection.url)) {
                const url = this.toUrl(mapping, reflection);
                urls.push(new typedoc_1.UrlMapping(url, reflection, mapping.template));
                reflection.url = url;
                reflection.hasOwnDocument = true;
            }
            for (const child of reflection.children || []) {
                if (mapping.isLeaf) {
                    this.applyAnchorUrl(child, reflection);
                }
                else {
                    this.buildUrls(child, urls);
                }
            }
        }
        else if (reflection.parent) {
            this.applyAnchorUrl(reflection, reflection.parent);
        }
        return urls;
    }
    toUrl(mapping, reflection) {
        return mapping.directory + '/' + this.getUrl(reflection) + '.md';
    }
    getUrl(reflection, relative, separator = '.') {
        let url = reflection.getAlias();
        if (reflection.parent &&
            reflection.parent !== relative &&
            !(reflection.parent instanceof typedoc_1.ProjectReflection)) {
            url =
                this.getUrl(reflection.parent, relative, separator) + separator + url;
        }
        return url;
    }
    applyAnchorUrl(reflection, container) {
        if (!reflection.url || !MarkdownTheme.URL_PREFIX.test(reflection.url)) {
            const reflectionId = reflection.name.toLowerCase();
            const anchor = this.toAnchorRef(reflectionId);
            reflection.url = container.url + '#' + anchor;
            reflection.anchor = anchor;
            reflection.hasOwnDocument = false;
        }
        reflection.traverse((child) => {
            if (child instanceof typedoc_1.DeclarationReflection) {
                this.applyAnchorUrl(child, container);
            }
        });
    }
    toAnchorRef(reflectionId) {
        return reflectionId;
    }
    getEntryPoint(project) {
        const entryPoint = this.owner.entryPoint;
        if (entryPoint) {
            const reflection = project.getChildByName(entryPoint);
            if (reflection) {
                if (reflection instanceof typedoc_1.ContainerReflection) {
                    return reflection;
                }
                else {
                    this.application.logger.warn('The given entry point `%s` is not a container.', entryPoint);
                }
            }
            else {
                this.application.logger.warn('The entry point `%s` could not be found.', entryPoint);
            }
        }
        return project;
    }
    getNavigation(project) {
        var _a, _b, _c;
        const buildNavigationGroups = (navigation, groups) => {
            groups
                .filter((group) => group.allChildrenHaveOwnDocument())
                .forEach((reflectionGroup) => {
                var _a, _b;
                let reflectionGroupItem = (_a = navigation.children) === null || _a === void 0 ? void 0 : _a.find((child) => child.title === reflectionGroup.title);
                if (!reflectionGroupItem) {
                    reflectionGroupItem = createNavigationItem(reflectionGroup.title);
                    (_b = navigation.children) === null || _b === void 0 ? void 0 : _b.push(reflectionGroupItem);
                }
                reflectionGroup.children.forEach((reflectionGroupChild) => {
                    var _a;
                    const reflectionGroupChildItem = createNavigationItem(reflectionGroupChild.getFullName(), reflectionGroupChild.url);
                    if (reflectionGroupItem) {
                        (_a = reflectionGroupItem.children) === null || _a === void 0 ? void 0 : _a.push(reflectionGroupChildItem);
                    }
                    const reflection = reflectionGroupChild;
                    if (reflection.groups) {
                        buildNavigationGroups(navigation, reflection.groups);
                    }
                });
            });
        };
        const createNavigationItem = (title, url) => {
            const navigationItem = new typedoc_1.NavigationItem(title.replace(/\"/g, ''), url);
            navigationItem.isLabel = !url;
            navigationItem.children = [];
            delete navigationItem.reflection;
            delete navigationItem.parent;
            return navigationItem;
        };
        const sortNavigationGroups = (a, b) => {
            const weights = {
                ['Namespaces']: 1,
                ['Enumerations']: 2,
                ['Classes']: 3,
                ['Interfaces']: 4,
                ['Type aliases']: 5,
                ['Variables']: 6,
                ['Functions']: 7,
                ['Object literals']: 8,
            };
            const aWeight = weights[a.title] || 0;
            const bWeight = weights[b.title] || 0;
            return aWeight - bWeight;
        };
        const entryPoint = this.getEntryPoint(project);
        const hasSeperateGlobals = this.readme !== 'none';
        const navigation = createNavigationItem(project.name);
        (_a = navigation.children) === null || _a === void 0 ? void 0 : _a.push(createNavigationItem(hasSeperateGlobals ? 'README' : 'Globals', this.entryFile));
        if (hasSeperateGlobals) {
            (_b = navigation.children) === null || _b === void 0 ? void 0 : _b.push(createNavigationItem('Globals', this.globalsFile));
        }
        if (entryPoint.groups) {
            buildNavigationGroups(navigation, entryPoint.groups);
        }
        (_c = navigation.children) === null || _c === void 0 ? void 0 : _c.sort(sortNavigationGroups);
        return navigation;
    }
    onPageEnd(page) {
        page.contents = page.contents
            ? MarkdownTheme.formatContents(page.contents)
            : '';
    }
    get mappings() {
        return [
            {
                kind: [models_1.ReflectionKind.Class],
                isLeaf: false,
                directory: 'classes',
                template: 'reflection.hbs',
            },
            {
                kind: [models_1.ReflectionKind.Interface],
                isLeaf: false,
                directory: 'interfaces',
                template: 'reflection.hbs',
            },
            {
                kind: [models_1.ReflectionKind.Enum],
                isLeaf: false,
                directory: 'enums',
                template: 'reflection.hbs',
            },
            {
                kind: [models_1.ReflectionKind.Namespace, models_1.ReflectionKind.Module],
                isLeaf: false,
                directory: 'modules',
                template: 'reflection.hbs',
            },
            ...(this.allReflectionsHaveOwnDocument
                ? [
                    {
                        kind: [models_1.ReflectionKind.Variable],
                        isLeaf: true,
                        directory: 'variables',
                        template: 'reflection.member.hbs',
                    },
                    {
                        kind: [models_1.ReflectionKind.TypeAlias],
                        isLeaf: true,
                        directory: 'types',
                        template: 'reflection.member.hbs',
                    },
                    {
                        kind: [models_1.ReflectionKind.Function],
                        isLeaf: true,
                        directory: 'functions',
                        template: 'reflection.member.hbs',
                    },
                    {
                        kind: [models_1.ReflectionKind.ObjectLiteral],
                        isLeaf: true,
                        directory: 'literals',
                        template: 'reflection.member.hbs',
                    },
                ]
                : []),
        ];
    }
    get entryFile() {
        return 'README.md';
    }
    get globalsFile() {
        return 'globals.md';
    }
    get hasSidebar() {
        return false;
    }
}
MarkdownTheme.HANDLEBARS = Handlebars.create();
MarkdownTheme.URL_PREFIX = /^(http|ftp)s?:\/\//;
__decorate([
    typedoc_1.BindOption('readme')
], MarkdownTheme.prototype, "readme", void 0);
__decorate([
    typedoc_1.BindOption('allReflectionsHaveOwnDocument')
], MarkdownTheme.prototype, "allReflectionsHaveOwnDocument", void 0);
exports.default = MarkdownTheme;
