// read-text-file/dist/read-text-file.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var iconv = require("iconv-lite");
var jschardet = require("jschardet");
function read(path) {
    return __awaiter(this, void 0, void 0, function () {
        var stat, fd, result, buffer;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, fileStat(path)];
                case 1:
                    stat = _a.sent();
                    fd = null;
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, , 5, 8]);
                    return [4 /*yield*/, openFile(path, "r")];
                case 3:
                    fd = _a.sent();
                    buffer = new Buffer(stat.size);
                    return [4 /*yield*/, readFile(fd, buffer, 0, stat.size, 0)];
                case 4:
                    _a.sent();
                    result = decode(buffer);
                    return [3 /*break*/, 8];
                case 5:
                    if (!((fd !== null) && (fd !== undefined))) return [3 /*break*/, 7];
                    return [4 /*yield*/, closeFile(fd)];
                case 6:
                    _a.sent();
                    _a.label = 7;
                case 7: return [7 /*endfinally*/];
                case 8: return [2 /*return*/, result];
            }
        });
    });
}
exports.read = read;
function readSync(path) {
    var stat = fs.statSync(path);
    var fd = null;
    var result;
    try {
        fd = fs.openSync(path, "r");
        var buffer = new Buffer(stat.size);
        fs.readSync(fd, buffer, 0, stat.size, 0);
        result = decode(buffer);
    }
    finally {
        if ((fd !== null) && (fd !== undefined)) {
            fs.closeSync(fd);
        }
    }
    return result;
}
exports.readSync = readSync;
function decode(buffer) {
    // TODO: fallback for when confidence is too low? (pass it as "defaultEncoding" below)
    // TODO: this is decoding the whole file twice (once to get encoding name, then again to really decode... should just take a portion of it to get the encoding name)
    var encodingName = getEncodingName(buffer);
    return iconv.decode(buffer, encodingName, { stripBOM: true, addBOM: false, defaultEncoding: "utf-8" });
}
function getEncodingName(buffer) {
    // TODO: set min confidence?
    var result = jschardet.detect(buffer);
    return result.encoding;
}
// TODO: share these, or try fs-promise (or similar)
function fileStat(path) {
    return new Promise(function (resolve, reject) {
        fs.stat(path, function (err, stats) {
            if ((null !== err) && (undefined !== err)) {
                reject(err);
            }
            else {
                resolve(stats);
            }
        });
    });
}
function openFile(path, flags) {
    return new Promise(function (resolve, reject) {
        fs.open(path, flags, function (err, fd) {
            if ((null !== err) && (undefined !== err)) {
                reject(err);
            }
            else {
                resolve(fd);
            }
        });
    });
}
function readFile(fd, buffer, offset, length, position) {
    return new Promise(function (resolve, reject) {
        fs.read(fd, buffer, offset, length, position, function (err, bytesRead, buffer) {
            if ((null !== err) && (undefined !== err)) {
                reject(err);
            }
            else {
                resolve({ bytesRead: bytesRead, buffer: buffer });
            }
        });
    });
}
function closeFile(fd) {
    return new Promise(function (resolve, reject) {
        fs.close(fd, function (err) {
            if ((null !== err) && (undefined !== err)) {
                reject(err);
            }
            else {
                resolve();
            }
        });
    });
}
